DEPLOY.md

```diff
@@ -0,0 +1,104 @@
+# Deployment
+
+This repository contains two services: the RAP Agent and the RAP
+Controller. Both are deployed automatically on merge to main.
+
+
+## Agent deployment
+
+This runs within the secure backends (one instance per backend) and is
+managed by the [backend-server](https://github.com/opensafely-core/backend-server/) repo.
+
+A timer checks for new builds of the Docker image and pulls them down,
+so there may be a short delay (ten minutes maybe?) between the Github
+Actions workflows completing and the new image being deployed.
+
+
+## Controller deployment
+
+We run a single instance of the controller which all the agents connect
+to. This is deployed as a Dokku app and at the time of writing runs on
+`dokku4`. A Github Actions workflow connects to the host and runs the
+dokku deploy process with the newly built image.
+
+Below are the steps used to configure the app initially (though the
+exact config should be regarded as illustrative and may well have
+changed by the time of reading).
+
+Assume we have a server correctly configured by the [sysadmin](https://github.com/bennettoxford/sysadmin)
+tooling. This will ensure that Dokku is installed and configured with the
+necessary plugins and that a `rap-controller` user has been created.
+
+First we create the app and do some basic configuration:
+```bash
+dokku apps:create rap-controller
+
+# Set up the domain
+dokku domains:set rap-controller controller.opensafely.org
+dokku letsencrypt:enable rap-controller
+
+# Disable zero-dowtime deploys for service, we don't ever want two of
+# these loops running simultaneously
+dokku checks:disable rap-controller service
+
+# Create an appropriately owned storage directory and mount it
+sudo mkdir -p /var/lib/dokku/data/storage/rap-controller
+sudo chown rap-controller:rap-controller /var/lib/dokku/data/storage/rap-controller
+dokku storage:mount rap-controller /var/lib/dokku/data/storage/rap-controller:/storage
```

Comment:
The ansible script in sysadmin will do this [here](https://github.com/bennettoxford/sysadmin/blob/main/infra/dokku_app_users.yaml#L14)

In fact, I'm not sure that the ansible script will work until the dokku app is created, because it runs dokku commands that expect the app

Reply:
Ah, so I had the same worry initially then I saw it ran with `check_mode: false` and assumed that meant it would just ignore failures. But now I've double-checked and that isn't what it does at all!

So yeah, this is a problem. But I don't understand how this is supposed to work because you can't create the app until the server is set up with Dokku, which is what Ansible is supposed to do. Maybe this has never been run from scratch against a fresh server?

I think updating the command so that it just outputs nothing if the app isn't configured yet would be the right thing to do here.

Thanks for catching this.

Reply:
I don't think it has ever been run from scratch. I think we switched all apps to run as non root a while ago and it was maybe set up then, but all the apps already existed

Reply:
There is an ignore-failure ansible option (can't remember exactly what it's called), maybe we just need that?